activation/e2e/poet_client.go

```diff
@@ -0,0 +1,151 @@
+package activation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/spacemeshos/merkle-tree"
+	"github.com/spacemeshos/merkle-tree/cache"
+	"github.com/spacemeshos/merkle-tree/cache/readwriters"
+	"github.com/spacemeshos/poet/hash"
+	"github.com/spacemeshos/poet/shared"
+
+	"github.com/spacemeshos/go-spacemesh/activation"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+type TestPoet struct {
+	mu    sync.Mutex
+	round int
+
+	expectedMembers int
+	registrations   chan []byte
+}
+
+func NewTestPoetBackend(expectedMembers int) *TestPoet {
```

Comment:
I feel like this should just be a generated gomock.

The `Submit` and `Proof` methods can be defined as functions/closures that are passed as `EXPECTED` to the generated mock.

Wdyt?

Reply:
I'm not sure. The logic is quite complex, it needs to build a valid poet proof. But the real problem is that `Submit()` and `Proof()` share state (i.e registrations and the round id). Using a generated mock would still require to have a struct with at least these two methods (and call them in the mock EXPECTs...). IMHO, it would only litter the code and make tests more difficult to write correctly w/o a benefit.

Reply:
I think we should use a slice instead of a chan for the registrations because otherwise a test will get stuck if fewer submits happen than expected. Also if more happen then expected the registrations will just be added to the wrong round?

I was thinking of somethign like this:

```go
func NewTestPoetClient(ctrl *gomock.Controller, expectedMembers, expectedRounds int) *activation.MockPoetClient {
	poet := activation.NewMockPoetClient(ctrl)
	poet.EXPECT().Id().AnyTimes().Return([]byte("http://poet.test"))
	poet.EXPECT().Address().AnyTimes().Return("http://poet.test")
	poet.EXPECT().PowParams(gomock.Any()).AnyTimes().Return(&activation.PoetPowParams{}, nil)

	var mu sync.Mutex
	var mockRound int
	registrations := make([][]byte, 0, expectedMembers)
	proofs := make(map[string]*types.PoetProofMessage, expectedRounds)

	var submitCall *gomock.Call
	var proofCall *gomock.Call

	for range expectedRounds {
		submitCall = poet.EXPECT().Submit(
			gomock.Any(),
			gomock.Any(),
			gomock.Any(),
			gomock.Cond(func(x any) bool { return len(x.([]byte)) == 32 }),
			gomock.Any(),
			gomock.Any(),
			gomock.Any(),
		).DoAndReturn(func(
			_ context.Context,
			_ time.Time,
			_, challenge []byte,
			_ types.EdSignature,
			nodeID types.NodeID,
			_ activation.PoetAuth,
		) (*types.PoetRound, error) {
			mu.Lock()
			defer mu.Unlock()
			registrations = append(registrations, challenge)
			return &types.PoetRound{ID: strconv.Itoa(mockRound), End: time.Now().Add(3 * time.Second)}, nil
		}).Times(expectedMembers)

		proofCall = poet.EXPECT().Proof(
			gomock.Any(),
			gomock.Cond(func(x any) bool { return x.(string) == strconv.Itoa(mockRound) }),
		).DoAndReturn(func(ctx context.Context, roundID string) (*types.PoetProofMessage, []types.Hash32, error) {
			mu.Lock()
			defer mu.Unlock()
			if proof, ok := proofs[roundID]; ok {
				return proof, nil, nil
			}

			mtree, err := merkle.NewTreeBuilder().WithHashFunc(shared.HashMembershipTreeNode).Build()
			if err != nil {
				return nil, nil, err
			}

			var members []types.Hash32
			for _, member := range registrations {
				if err := mtree.AddLeaf(member[:]); err != nil {
					return nil, nil, err
				}
				members = append(members, types.Hash32(member))
			}
			registrations = registrations[:0]

			challenge := mtree.Root()

			const leaves = uint64(shared.T)

			treeCache := cache.NewWriter(
				func(uint) bool { return true },
				func(uint) (cache.LayerReadWriter, error) { return &readwriters.SliceReadWriter{}, nil },
			)

			labelHashFunc := hash.GenLabelHashFunc(challenge)
			merkleHashFunc := hash.GenMerkleHashFunc(challenge)
			tree, err := merkle.NewTreeBuilder().WithHashFunc(merkleHashFunc).WithCacheWriter(treeCache).Build()
			if err != nil {
				return nil, nil, err
			}

			makeLabel := shared.MakeLabelFunc()
			for i := range leaves {
				parkedNodes := tree.GetParkedNodes(nil)
				err := tree.AddLeaf(makeLabel(labelHashFunc, i, parkedNodes))
				if err != nil {
					return nil, nil, err
				}
			}

			root := tree.Root()
			cacheReader, err := treeCache.GetReader()
			if err != nil {
				return nil, nil, err
			}
			provenLeafIndices := shared.FiatShamir(root, leaves, shared.T)
			_, provenLeaves, proofNodes, err := merkle.GenerateProof(provenLeafIndices, cacheReader)
			if err != nil {
				return nil, nil, err
			}

			proof := &types.PoetProofMessage{
				PoetProof: types.PoetProof{
					MerkleProof: shared.MerkleProof{
						Root:         root,
						ProvenLeaves: provenLeaves,
						ProofNodes:   proofNodes,
					},
					LeafCount: leaves,
				},
				RoundID:       roundID,
				PoetServiceID: []byte("http://poet.test"),
				Statement:     types.Hash32(challenge),
			}
			proofs[roundID] = proof
			mockRound++
			return proof, members, nil
		}).After(submitCall).MinTimes(1)
	}

	// block additional submit calls
	poet.EXPECT().Submit(
		gomock.Any(),
		gomock.Any(),
		gomock.Any(),
		gomock.Any(),
		gomock.Any(),
		gomock.Any(),
		gomock.Any(),
	).AnyTimes().DoAndReturn(func(
		ctx context.Context,
		_ time.Time,
		_, challenge []byte,
		_ types.EdSignature,
		_ types.NodeID,
		_ activation.PoetAuth,
	) (*types.PoetRound, error) {
		<-ctx.Done()
		return nil, ctx.Err()
	}).After(proofCall).AnyTimes()

	return poet
}
```

Reply:
This won't work. `Proof()` must wait for `expectedMembers` to register before it builds the proof. Otherwise, a client who called `Proof()`  before all others `Submit` would cause the proof to be generated too early and tests would fail. Adding 3 seconds to `End` time is not a good idea as it will be flaky on slower runners and again makes the test longer and must be in-sync with epoch duration and cycle gap configuration to make sense. In my solution, the proof is available immediately when all expected members register, giving the maximum possible time for the post proofvingto succeed.

Besides, IMHO the solution with auto-generated mock is unreadable.

Reply:
In the current solution if fewer than expected callers call submit the test will get stuck and if more than the expected call it, it will generate a faulty proof.

Also Proof isn't safe to be called concurrently or even possible to be called more than once for a given round.

It is a few seconds faster at the expense of being racy :shrug: also if proof is called before the expected number of submits in my suggested solution the test will fail with a message saying exactly that this happened...

Reply:
Alternatively you can do similar changes to the current approach where you replace the registration chan with a slice and an additional "ready" chan for each round that the proof method blocks on to achieve a similar effect. 

This would save the 3 seconds that my suggested solution is slower.

You could also add a check to the submit method that fails the test when adding the registration to the slice would cause it to become bigger than expected.

Reply:
> In the current solution if fewer than expected callers call submit the test will get stuck

I can fix it to bail out when the context is canceled (and the context used by nipostBuilder has a timeout).

> and if more than the expected call it, it will generate a faulty proof.

The proof wouldn't be faulty, but it wouldn't contain the extra members. Such a condition is a test error. The contract here is "you configure me to generate proofs with 10 members - I will create proofs with exactly 10 members". I can add a check to fail in this scenario but it's unnecessary in my opinion.

>Also Proof isn't safe to be called concurrently or even possible to be called more than once for a given round.
>
> It is a few seconds faster at the expense of being racy

Yes, it's not safe and that's documented. Calls to Proof must be, and are synchronized by the `PoetService` so it is fine and it's not racy.

The proposed solution is time-dependent and thus 100% flaky and easy to break unintentionally.